subprojects/core/src/main/java/org/gradle/internal/scripts/BuildScriptCompileUnitOfWork.java

```diff
@@ -104,12 +125,43 @@ public InputFingerprinter getInputFingerprinter() {
 
     @Override
     public void visitOutputs(File workspace, OutputVisitor visitor) {
-        File classesDir = classesDir(workspace);
-        OutputFileValueSupplier outputValueSupplier = OutputFileValueSupplier.fromStatic(classesDir, fileCollectionFactory.fixed(classesDir));
-        visitor.visitOutputProperty("classesDir", TreeType.DIRECTORY, outputValueSupplier);
+        File originalDir = originalDir(workspace);
+        OutputFileValueSupplier originalDirValue = OutputFileValueSupplier.fromStatic(originalDir, fileCollectionFactory.fixed(originalDir));
+        visitor.visitOutputProperty("originalDir", TreeType.DIRECTORY, originalDirValue);
+
+        File transformedDir = transformedDir(workspace);
+        OutputFileValueSupplier transformedDirValue = OutputFileValueSupplier.fromStatic(transformedDir, fileCollectionFactory.fixed(transformedDir));
+        visitor.visitOutputProperty("transformedDir", TreeType.DIRECTORY, transformedDirValue);
+    }
+
+    private static File originalDir(File workspace) {
+        return new File(workspace, "original");
+    }
+
+    private static File transformedDir(File workspace) {
+        return new File(workspace, "transformed");
+    }
+
+    private static File transformedJar(File workspace) {
+        return new File(transformedDir(workspace), TRANSFORMED_BUILD_SCRIPT_JAR);
     }
 
-    private static File classesDir(File workspace) {
-        return new File(workspace, "classes");
+    public static class BuildScriptCompilationOutput {
+
+        private final File originalDir;
+        private final File output;
+
+        public BuildScriptCompilationOutput(File originalDir, File output) {
+            this.originalDir = originalDir;
+            this.output = output;
+        }
+
+        public File getOutput() {
+            return output;
+        }
+
+        public File getOriginalDir() {
+            return originalDir;
+        }
```

Comment:
❓ So `originalDir` is the compiled `.class` files _without_ instrumentation, right? And `output` is the same `.class` files, but _with_ instrumentation? If so, I think that's what we should call these methods.

I'm a bit confused why we need to carry both.

Could we make this clearer if compilation and instrumentation were separated? It feels like they shouldn't be handled in one place.

Reply:
I refactored code a bit and made it a bit more obvious. Now only Groovy's Unit of work  returns some additional data, that is more explicit: it returns metadata and instrumented jar, there is no more original dir.